Query Optimization

andremanzo1 opened the pull request:
# Description
Applied the suggestions by @simonbtomlinson from [postgresQueryOptimization.md](https://github.com/OpenEnergyDashboard/DesignDocs/blob/main/postgresQueryOptimization/postgresQueryOptimization.md). Fixing the main issue of slow execution times of meter_3d_readings_unit.

## Type of change

- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [ ] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

The issue with bar in Postgres optimization is not addressed here.

andremanzo1 committed:
created a index for hourly_readings_unit. Removed redundant if statement and replaced it with the else if. Added a new constraint at line 141 as suggested from simon

andremanzo1 committed:
added commanet explaining the index: By indexing both columns together, the database can efficiently handle queries that involve both meter_id and time_interval

andremanzo1 committed:
added create_function_get_3d_readings.sql to migrations. OED version 1.0.0 - 2.0.0

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -9,6 +9,10 @@ not exceed the start/end times for the readings in the supplied meter. This can
 because infinity is used to indicate to graph all readings. This version does it to the nearest
 day by using the day reading view and is used by 3D readings which only allow days and a single meter.
  */
```
A few items:

- The comment above was for the shrink_tsrange_to_meter_readings_by_day function below. The index you inserted should not be between them.
- There is an extra space at the start of the line on this line and on the CREATE line below.
- I'm wondering how this relates to the idx_daily_readings_unit index in src/server/sql/reading/create_reading_views.sql. That one is on the interval. Is it still needed for other queries? It notes it does not do hourly (not needed for 3D) but I wonder if hourly should be added or it is not needed at all. Something to think through and test for timing.

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -138,7 +126,7 @@ BEGIN
                 FROM (
                     SELECT hour
-                    FROM generate_series(
+                    FROM generate_series( 
```
Please remove extra blank space at end of line that was added.

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -151,6 +139,7 @@ BEGIN
                 -- Only want readings that lie within this slice of the desired data
                 AND lower(hr.time_interval) >= hours.hour
                 AND upper(hr.time_interval) <= hours.hour + reading_length_interval
+                AND lower(hr.time_interval) <= hours.hour + reading_length_interval
```
It would be nice to add a comment on why the new line of SQL is there.

huss commented on src/server/migrations/1.0.0-2.0.0/sql/readings/create_function_get_3d_readings.sql:
```diff
@@ -0,0 +1,205 @@
+ CREATE INDEX IF NOT EXISTS idx_hourly_readings_unit_meter_time
+ON hourly_readings_unit (meter_id, lower(time_interval));
+
+CREATE OR REPLACE FUNCTION shrink_tsrange_to_meter_readings_by_day(tsrange_to_shrink TSRANGE, meter_id_desired INTEGER)
```
This function and group_3d_readings_unit do not seem to be modified by your changes. If so, they should not be in the migration since there is no change that needs to be done.

huss reviewed (changes requested):
Thanks to @andremanzo1 for their first contribution to OED. The check box to indicate that everyone who worked on this is not checked. That needs to be done. If anyone else worked on this then their GitHub IDs should be listed in the description and they need to sign the CLA.

There is a merge conflict that is blocking this pull request. Development needs to be merged so it is up-to-date and the conflict carefully resolved.

I've made a few other comments within the code to address. Please let me know if you want to discuss any of them.

I timed getting meters for 3D and see similar results to your timings so that looks good. I was curious about group and 3D so ran that too. It seems to add almost 100 ms to the execution time (on my machine). That is more than what I might have expected since it uses the 3D meter code to do the main work. I know this is asking for more work but it would be nice to do timings on your side, document, include a full explain to see what is going on and see if we can figure out why this is the case (if you see the same result). I used the amp meters and groups to do my testing. I'm happy to discuss this.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/andremanzo1/1450

andremanzo1 committed:
removed functions, added suggestion at line 168 for implementing a view that caluclates the sum and the readings start and end timestamp

andremanzo1 committed:
updated file

andremanzo1 committed:
removed functions that were not modified. Only keeping meter_3d_readings_unit

andremanzo1 committed:
adjusted spacing

huss commented:
> If anyone else worked on this then their GitHub IDs should be listed in the description and they need to sign the CLA.

The author verified that only their name was needed.

> There is a merge conflict that is blocking this pull request. Development needs to be merged so it is up-to-date and the conflict carefully resolved.

I pushed a commit to resolve this.

> I timed getting meters for 3D and see similar results to your timings so that looks good. I was curious about group and 3D so ran that too. It seems to add almost 100 ms to the execution time (on my machine). That is more than what I might have expected since it uses the 3D meter code to do the main work. I know this is asking for more work but it would be nice to do timings on your side, document, include a full explain to see what is going on and see if we can figure out why this is the case (if you see the same result). I used the amp meters and groups to do my testing. I'm happy to discuss this.

After talking with @andremanzo1 and thinking about the complications, it was decided that this would be a separate issue. It is now described in https://github.com/OpenEnergyDashboard/DesignDocs/blob/main/views/views.md.

huss committed:
Misc fixes

- Formatting/blank lines/spaces.
- Remove unneeded line.
- Fix up a couple of comments.

huss reviewed (approved):
Thanks to @andremanzo1 for the updates. I updated a few things and all comments are resolved. Everything seems to run fine and timing looks good. Congratulations on your first accepted contribution to OED.
